crates/relayer/src/chain/ckb4ibc/utils.rs

```diff
@@ -467,3 +511,101 @@ pub fn transaction_to_event(
     };
     Ok(event)
 }
+
+pub fn get_ibc_merkle_proof(height: Height, encoded: Vec<u8>) -> Result<Proofs, Error> {
+    let useless_client_proof = vec![0u8].try_into().unwrap();
+    let useless_consensus_proof =
+        ConsensusProof::new(vec![0u8].try_into().unwrap(), Height::default()).unwrap();
+    Proofs::new(
+        encoded.try_into().unwrap(),
+        Some(useless_client_proof),
+        Some(useless_consensus_proof),
+        None,
+        height,
+    )
+    .map_err(|err| Error::other_error(err.to_string()))
+}
+
+#[derive(EthAbiCodec, EthAbiType)]
+struct AxonObjectProof {
+    pub ckb_transaction: Vec<u8>,
+    pub block_hash: [u8; 32],
+    pub proof_payload: VerifyProofPayload,
+}
+
+pub async fn generate_tx_proof_from_block(
+    rpc_client: &impl CkbReader,
+    tx_hash: &H256,
+) -> Result<Option<Proofs>, Error> {
+    let block_hash = rpc_client
+        .get_transaction(tx_hash)
+        .await?
+        .map(|v| v.tx_status.block_hash);
+    let Some(Some(block_hash)) = block_hash else {
+        return Err(Error::other_error(format!(
+            "cannot find block_hash from tx {}",
+            hex::encode(tx_hash)
+        )));
+    };
+
+    // collect transaction hashes from block
+    let mut transaction: Option<Transaction> = None;
+    let block = rpc_client.get_block(&block_hash).await?;
+    let tx_hashes = block
+        .transactions
+        .iter()
+        .map(|tx| {
+            if &tx.hash == tx_hash {
+                transaction = Some(tx.inner.clone().into());
+            }
+            tx.hash.clone()
+        })
+        .collect_vec();
+    let witness_hashes = block
+        .transactions
+        .into_iter()
+        .map(|tx| Transaction::from(tx.inner).calc_witness_hash().unpack())
+        .collect_vec();
```

Comment:
Nit: it's not necessary to get all transactions to calculate the roots. You can calculate `raw_transactions_root` and `witnesses_root` from the proofs with e.g. [`MerkleProof::root`](https://docs.rs/merkle-cbt/0.3.2/merkle_cbt/merkle_tree/struct.MerkleProof.html#tymethod.root).

Reply:
I didn't get it, if I should get proof through method `rpc.get_transaction_and_witness_proof`, all transaction hashes in a block is needed, which I should call method `rpc.get_block` for getting a block

Reply:
You shouldn't need to get all leaves (tx witness hashes) and put them in the proof. It would defeat the purpose of merkle proofs, making the proof `O(n)` instead of `O(log n)` (where n is the number of leaves). My guess is you only need to pass the witness hash of the paricular transaction as `leaves`.

Reply:
check again